src/Data/ArrayBuffer/Typed.purs

```diff
@@ -1,87 +1,431 @@
+
+foldrWithIndex' :: forall a b t. TypedArray a t => (t -> b -> Offset -> b) -> b -> ArrayView a -> b
+foldrWithIndex' f i = unsafePerformEffect <<< foldrM (\x a o -> pure (f x a o)) i
+
+foldl1 :: forall a t. TypedArray a t => (t -> t -> t) -> ArrayView a -> t
+foldl1 f = foldl1WithIndex (\_ a x -> f a x)
+
+foldl1WithIndex :: forall a t. TypedArray a t => (Offset -> t -> t -> t) -> ArrayView a -> t
+foldl1WithIndex f = unsafePerformEffect <<< foldl1M (\acc x o -> pure (f o acc x))
+
+foldr1 :: forall a t. TypedArray a t => (t -> t -> t) -> ArrayView a -> t
+foldr1 f = foldr1WithIndex (\_ a x -> f a x)
+
+foldr1WithIndex :: forall a t. TypedArray a t => (Offset -> t -> t -> t) -> ArrayView a -> t
+foldr1WithIndex f = unsafePerformEffect <<< foldr1M (\x a o -> pure (f o x a))
+
+foreign import copyWithinImpl :: forall a. EffectFn4 (ArrayView a) Offset Offset (Nullable Offset) Unit
+
+-- | Internally copy values - see [MDN's spec](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/copyWithin) for details.
+copyWithin :: forall a. ArrayView a -> Offset -> Offset -> Maybe Offset -> Effect Unit
+copyWithin a t s me = runEffectFn4 copyWithinImpl a t s (toNullable me)
+
+foreign import reverseImpl :: forall a. EffectFn1 (ArrayView a) Unit
+
+-- | Reverses a typed array in-place.
+reverse :: forall a. ArrayView a -> Effect Unit
+reverse a = runEffectFn1 reverseImpl a
+
+foreign import setImpl :: forall a b. EffectFn3 (ArrayView a) Offset b Unit
+
+setInternal :: forall a b. (b -> Length) -> ArrayView a -> Maybe Offset -> b -> Effect Boolean
+setInternal lenfn a mo b =
+  let o = fromMaybe 0 mo
+  in if o >= 0 && lenfn b <= length a - o
+     then runEffectFn3 setImpl a o b *> pure true
+     else pure false
+
+
+-- | Stores multiple values in the typed array, reading input values from the second typed array.
+setTyped :: forall a. ArrayView a -> Maybe Offset -> ArrayView a -> Effect Boolean
+setTyped = setInternal length
+
+
+-- | Copy the entire contents of the typed array into a new buffer.
+foreign import sliceImpl :: forall a. Fn3 (ArrayView a) (Nullable Offset) (Nullable Offset) (ArrayView a)
+
+-- | Copy part of the contents of a typed array into a new buffer, between some start and end indices.
+slice :: forall a. ArrayView a -> Range -> ArrayView a
+slice a mz = case mz of
+  Nothing -> runFn3 sliceImpl a null null
+  Just (Tuple s me) -> runFn3 sliceImpl a (notNull s) (toNullable me)
+
+foreign import sortImpl :: forall a. EffectFn1 (ArrayView a) Unit
+
+-- | Sorts the values in-place
+sort :: forall a. ArrayView a -> Effect Unit
+sort a = runEffectFn1 sortImpl a
+
+
+foreign import subArrayImpl :: forall a. Fn3 (ArrayView a) (Nullable Offset) (Nullable Offset) (ArrayView a)
+
+-- | Returns a new typed array view of the same buffer, beginning at the index and ending at the second.
+-- |
+-- | **Note**: there is really peculiar behavior with `subArray` - if the first offset argument is omitted, or
```

Comment:
Is this behaviour for `subarray` documented somewhere?

Reply:
I wish. I can't find it in the ECMA spec, but it's pretty obvious if you try it in psci. I'm not sure how to find a doc on this either